design/proposals/spike-cli-dockerless.md

```diff
@@ -0,0 +1,257 @@
+
+#### Deployment of Go CLI
+
+The Go CLI is compiled down to a single executable binary. Go gives us the [ability to cross-build executables](https://www.digitalocean.com/community/tutorials/how-to-build-go-executables-for-multiple-platforms-on-ubuntu-16-04) so we don't need to hop to specific platform to compile the code. 
+
+#### Deployment of Python CLI
+
+The Python CLI is [single executable](https://github.com/cyberark/conjur-api-python3/releases/tag/v0.0.5), packed with the required runtime and dependencies. For this option there is no need for Python or other prerequisites to be installed on machine. Each executable will need to be bundled on each supported platform. 
+
+***NOTE***: Many platforms (except for Linux) require signed executables so might be an extra step in the deployment process
+
+|            | Testing                                                      |
+| ---------- | ------------------------------------------------------------ |
+| Ruby CLI   | Tested very well                                             |
+| Go CLI     | Pipeline maintained but the testing is not at the level it needs to be. Existing functionality will demand additional testing. <br />Considering the limited implemented functionality, new tests will need to be added if we choose to expand. |
+| Python CLI | Tested very well                                             |
+
+None of our CLIs test against other platforms besides Linux.
+
+|            | Security |
+| ---------- | -------- |
+| Ruby CLI   | See scan |
+| Go CLI     | See scan |
+| Python CLI | See scan |
+
+#### CLI commands to be supported
+
+The CLI should support the following commands:
+
+*Core* -  `init`, `authn` (`login`, `authenticate`),  `policy (load)`, `list`, `show `, `rotate_api_key` (`host` / `user`) and `fetch secrets`
+
+*Additional we might want to consider* - `resource`, `role`, `check`, `help`
+
+#### Estimation
+
+The following table summarizes the findings criteria detailed above and translates them to numbered equivalents. 
+
+1 - the least work
+
+3 - the most work
+
+|            | Functionality | Quality | UX   | Deployment | Testing | Security |
+| ---------- | ------------- | ------- | ---- | ---------- | ------- | -------- |
+| Ruby CLI   | 1             | 1       | 2    | 3          | 1       | See scan |
+| Go CLI     | 3             | 3       | 3    | 2          | 3       | See scan |
+| Python CLI | 2             | 2       | 1    | 1          | 2       | See scan |
+
+Ruby CLI ~11 weeks
+
+- Ramp-up on creating packages and create design for CLI (~1 week)
+- Provide package (rpm, pkg, deb, etc) / [deliver as executable](https://www.ruby-toolbox.com/categories/packaging_to_executables) (CLI + dependencies) (~3 week)
+- Expand current testing if needed (~1 week)
+- Build infrastructure for testing on new platforms (Windows, macOS, REHL environments) (~3 weeks)
+- Run a few CLI commands on new supported platforms (Windows, macOS environments) (~2 weeks)
```

Comment:
If the CLI is intended for the Dockerless DAP effort, why would we do any additional work on new platforms?  Doesn't this expand the work beyond the immediate need (to provide a Conjur CLI for Dockerless DAP)?

Reply:
> If the CLI is intended for the Dockerless DAP effort

Do you mean b/c Dockerless DAP is supposed to be initially deployed on REHL?

This is a good point! I cannot speak to requirements and scope but I would imagine the answer would be because how a customer will interface with DAP will change as a result of this decision. The ultimate goal is to choose the main CLI the customer will use to interface with Conjur/DAP and _**if**_ needed, ultimately swap the Ruby CLI for that one (Python/Go). If the CLI is swapped then we will want to support the chosen one across all our main platforms because is now a main component.

@InbalZilberman please feel free to chime in here because this is a question of scope

Reply:
@jvanderhoof the platform that DAP runs on is not necessarily the same platform that users of the CLI will be running from. personally I always run the CLI from my local machine, which is a Mac. I think this requirement is intended to enable the same capability for customers using Dockerless DAP (ability to run the CLI from their local machine).